### PR TITLE
SONARJAVA-1960: flow reporting for S2222

### PR DESCRIPTION
**Pull request by saberduck**
S2222 should rely on flow reporting rather than syntax node in constraint

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java [hunk 1]
```diff
@@ -103,10 +108,20 @@ public void reportIssue(Tree tree, String message, Set<List<JavaFileScannerConte
...
         continue;
       }
-      finalNode.getLearnedConstraints().stream()
-        .filter(lc -> lc.getSv().equals(currentVal) && constraintPredicate.test(lc.getConstraint()))
-        .findFirst()
-        .ifPresent(lc -> flow.add(new JavaFileScannerContext.Location("...", finalNode.parent().programPoint.syntaxTree())));
+
+      if (finalNode.getLearnedConstraints().stream()
+        .filter(lc -> lc.getSv().equals(currentVal))
+        .map(lc -> lc.getConstraint())
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java]: MINOR Replace this lambda with a method reference. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1612)

File: java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java [hunk 2]
```diff
@@ -103,10 +108,20 @@ public void reportIssue(Tree tree, String message, Set<List<JavaFileScannerConte
...
+        .anyMatch(addToFlow)) {
+        flow.add(new JavaFileScannerContext.Location("...", finalNode.parent().programPoint.syntaxTree()));
+      }
+      if (finalNode.getLearnedConstraints().stream()
+        .filter(lc -> lc.getSv().equals(currentVal))
+        .map(lc -> lc.getConstraint())
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java]: MINOR Replace this lambda with a method reference. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1612)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 1 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SECheck.java#L105](https://github.com/SonarSource/sonar-java/blob/79223e466b245d1d12e80c0b7b9349e4bbab1e4c/java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java#L105): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SECheck.java#L105](https://github.com/SonarSource/sonar-java/blob/dae7eb7ec972da3cb36ffd12db14171749a47604/java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java#L105): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [SECheck.java#L106](https://github.com/SonarSource/sonar-java/blob/c4363d8769a95137dd892c4a85f1c72c40373e19/java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java#L106): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS135)

**Comment by saberduck** [comment]
I will deal with the SonarQube issue in 1963

**Pull request merged**